internal/credential/vault/repository_ssh_certificate_credential_library.go

```diff
@@ -0,0 +1,344 @@
+	}
+	if l.Username == "" {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "no username")
+	}
+	l = l.clone()
+
+	if l.KeyType == "" {
+		l.KeyType = KeyTypeEd25519
+	}
+
+	if l.GetCredentialType() == "" {
+		l.SSHCertificateCredentialLibrary.CredentialType = string(credential.SshCertificateType)
+	}
+	if l.GetCredentialType() != string(credential.SshCertificateType) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "invalid credential type")
+	}
+
+	id, err := newSSHCertificateCredentialLibraryId()
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	l.setId(id)
+
+	oplogWrapper, err := r.kms.GetWrapper(ctx, projectId, kms.KeyPurposeOplog)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op, errors.WithMsg("unable to get oplog wrapper"))
+	}
+
+	var newSSHCertificateCredentialLibrary *SSHCertificateCredentialLibrary
+	_, err = r.writer.DoTx(ctx, db.StdRetryCnt, db.ExpBackoff{},
+		func(_ db.Reader, w db.Writer) error {
+			var msgs []*oplog.Message
+			ticket, err := w.GetTicket(ctx, l)
+			if err != nil {
+				return errors.Wrap(ctx, err, op, errors.WithMsg("unable to get ticket"))
+			}
+
+			// insert credential library
+			newSSHCertificateCredentialLibrary = l.clone()
+			var lOplogMsg oplog.Message
+			if err := w.Create(ctx, newSSHCertificateCredentialLibrary, db.NewOplogMsg(&lOplogMsg)); err != nil {
+				return errors.Wrap(ctx, err, op)
+			}
+			msgs = append(msgs, &lOplogMsg)
+
+			metadata := l.oplog(oplog.OpType_OP_TYPE_CREATE)
+			if err := w.WriteOplogEntryWith(ctx, oplogWrapper, ticket, metadata, msgs); err != nil {
+				return errors.Wrap(ctx, err, op, errors.WithMsg("unable to write oplog"))
+			}
+			return nil
+		},
+	)
+
+	if err != nil {
+		if errors.IsUniqueError(err) {
+			return nil, errors.Wrap(ctx, err, op, errors.WithMsg(fmt.Sprintf("in credential store: %s: name %s already exists", l.StoreId, l.Name)))
+		}
+		return nil, errors.Wrap(ctx, err, op, errors.WithMsg(fmt.Sprintf("in credential store: %s", l.StoreId)))
+	}
+	return newSSHCertificateCredentialLibrary, nil
+}
+
+// UpdateSSHCertificateCredentialLibrary updates the repository entry for l.PublicId with
+// the values in l for the fields listed in fieldMaskPaths. It returns a
+// new SSHCertificateCredentialLibrary containing the updated values and a count of the
+// number of records updated. l is not changed.
+//
+// l must contain a valid PublicId. Only Name, Description, VaultPath,
+// HttpMethod, HttpRequestBody, and MappingOverride can be updated. If
+// l.Name is set to a non-empty string, it must be unique within l.StoreId.
+//
+// An attribute of l will be set to NULL in the database if the attribute
+// in l is the zero value and it is included in fieldMaskPaths except for
+// HttpMethod.  If HttpMethod is in the fieldMaskPath but l.HttpMethod
+// is not set it will be set to the value "GET".  If storage has a value
+// for HttpRequestBody when l.HttpMethod is set to GET the update will fail.
```

Comment:
This comment needs updating since it references fields that are not a part of a SSHCerticiateCredentialLibrary.